[20208] New comment field for work package split/full screen view

HDinger opened the pull request:
This creates a new comment field for the work package split and full screen view.
Requirements:
- [x] Design after the visual (http://clients.zzmedia.net/openproject/workpackages/fullscreen/small/)
- [x] Initially the field is small and with text in it
- [x] On click the field grows
- [x] When something is written, the text disappears and a 'save'-button appears
- [x] The 'save'-button should look like on the other input fields and actually save the comment
- [x] When the text grows, the field grows, too (jQuery-autogrow-Plugin)
- [x] The box should appear above or below the rest depending on the sort direction

If https://github.com/opf/openproject/pull/3366 is merged before this PR. Otherwise, the feature needs to be added in https://github.com/opf/openproject/pull/3366 or as a separate PR
- [ ] The field shall be available in split and full screen

https://community.openproject.org/work_packages/20208

HDinger committed:
Editable comment field

This provides the basic functionality to deliver a comment field
in the in-place editing style.

Adds the `autogrow` plugin

This still breaks a number of features currently, such as formatted
previewing.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://hdinger-epic-20208-edit-comment-one-step-details-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://hdinger-epic-20208-edit-comment-one-step-details-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther commented:
I've added a basic fix for the broken toolbar sizing in IE10 / 11. There may be other issues still.

oliverguenther committed:
Remove replaced activity_comment

oliverguenther committed:
Correct comment behavior with optional notifications

This allows submitting comments with and without notification
for the created journal.

Using hyperagent resource on activities, the resource `addComment` is
replaced when executing `fetch` on the link.
This breaks subsequent comment requests. This is fixed by requesting
manually on the resource link using `$http`.

This commit also provides the functionality required to comment with or
without notification propagation, and changes the backend to comply to
the APIv3 specification, which defines comments to be sent as a JSON
`{ comment: { raw: 'my comment' } }`.

This commit also provides:

* Displaying the comment input on overview depending on sort setting
* Disable inplace controls when comment is empty
* correctly hiding the comment based on link availability

oliverguenther committed:
Correct height when expanding inplace textareas with autogrow

When using `.-autogrow` on textareas in inplace edit (e.g., description),
the height is not correctly set as the selector is not visible.

Only when starting to EDIT the textarea, the changes are being picked up
and height is set correctly.

This commit thus provides a second event listener to `focusin` events
that cause the textareas to resize properly.

oliverguenther commented:
Note that I'm still working on adapting the old `comment-activity` tests, thus travis will likely disagree with my changes for the time being.

oliverguenther commented on frontend/app/work_packages/directives/work-package-comment-directive.js:
```diff
@@ -0,0 +1,148 @@
+      return true;
+    };
+
+    ctrl.submit = function(notify) {
```
Ehh.. notifiy _is_ being used (l. 66)

oliverguenther committed:
Test for comment input

ulferts commented on frontend/app/templates/work_packages/tabs/overview.html:
```diff
@@ -57,6 +57,11 @@ <h3 class="attributes-group--header-text">{{ I18n.t('js.label_latest_activity')
   </div>
 
   <exclusive-edit>
+    <div ng-show="activitiesSortedInDescendingOrder === true">
```
I don't think that the `=== true` is necessary

oliverguenther committed:
Fix invalid ng-show on comments

With multiple `ng-show`, the latter element will be selected, even when its hidden.
This causes the quote button to (appear to) malfunction when sorting
descending.

HUGE props to @ulferts for finding that one.

oliverguenther committed:
Replace exclusive-edit with transcluded comment directive

This commit removes the somewhat superfluous exlusiveEditController
in favor of event-based passing of quoted text to the correct comment
field.

The comment directive now decides for itself whether it wants to be
placed above or below the activities and uses transclusion for that
purpose.

Editing comments is now no longer exlusive, but that shouldn't be
necessary anyway.

oliverguenther commented:
@ulferts Thanks for finding the quoting issue. I've addressed your concerns regarding the exlusive edit functionality. 

The latest commit replaces the exlusiveEditController with a simple event for passing of quoted text to the correct comment field.

The comment directive now decides for itself whether it wants to be
placed above or below the activities and uses transclusion for that
purpose.

Editing comments is now no longer exlusive, but that shouldn't be
necessary anyway.

oliverguenther committed:
Editing of comments with in-place button styles and preview

This commit changes the comment edit and cancel buttons to reflect the
in-place editing style.

Also allows previewing of edited comments and also shows a
notification-box when comments are edited.

oliverguenther committed:
Propagate and distribute field and comment submits in WP Field Service

This commit allows parallel saving of comments and fields, however
this currently results in a conflicting modification due to two saves
with thes same lock version.

oliverguenther committed:
Parallel sending of comment and fields in WP split view

Parallel submitting of comments alongside WP edits wasn't possible
natively due to the increased lock_version in work packages, caused by
`touch`ing and explicitly saving the work package in the process of
adding a journal.

This commit works around this issue by updating the timestamp columns
directly, bypassing optimistic locking in the work package table.

This allows comments to be submitted transparently with other fields to
the user, while actually, we send it in two subsequent steps.

oliverguenther committed:
Fix stale object from fixed lock_version in legacy spec

The affected test used the next lock_version, which now fails
as the journal no longer saves the container.

ulferts commented on app/assets/stylesheets/content/_components_add_comments_onclick.lsg:
```diff
@@ -1,10 +0,0 @@
-<textarea class="add-comment-text-big" placeholder="Add your comments here" rows=4></textarea>
-<button class="button"><i class="icon-yes icon-left"></i>Add comment</button><button class="button"><i class="icon-close icon-left"></i>Cancel</button>
-</div>
-
```
:+1: for cleaning up

ulferts commented on app/assets/javascripts/autogrow.js:
```diff
@@ -26,16 +26,16 @@
+  $(function() {
+    $('.-autogrow').autogrow({
+      onInitialize: true
+    });
```
Is it possible to always add a newline just like it can be done with https://github.com/monospaced/angular-elastic?

When typing a long text (not hitting enter before the wrap happens) the animation is not really smoth. This is probably because of the one thread limitation browsers have. Always having a newline ready should make it nicer to use.

Additionally, as a user, I don't know what will happen until I get past the wrap, so I might be tempted to increase the height unnecessarily.

ulferts commented on frontend/bower.json:
```diff
@@ -28,7 +28,8 @@
     "mousetrap": "~1.4.6",
-    "ng-file-upload": "~5.0.9"
+    "ng-file-upload": "~5.0.9",
+    "autogrow": "ultimatedelman/autogrow#master#a243e1a5cde9bffad11c978153c1b21a6bdc580a"
```
Why don't we use a released version here?

HDinger replied:
Because the latest version doesn't include the bower.json, but the actual commit does.

ulferts commented on app/assets/stylesheets/content/_in_place_editing.sass:
```diff
@@ -265,6 +267,29 @@ a.inplace-editing--trigger-link,
+  // Disabled comment submit styles when empty
+  .inplace-edit--control--save[disabled],
+  .inplace-edit--control--send[disabled]
+    background-color: $inplace-edit--bg-color--disabled
```
I don't think this rule should be specific to adding comments.

HDinger replied:
:ballot_box_with_check:

oliverguenther replied:
I'll move it, even though I can't address the styling for other non-empty fields in this PR.
:ballot_box_with_check:

ulferts commented on app/assets/stylesheets/content/_in_place_editing.sass:
```diff
@@ -265,6 +267,29 @@ a.inplace-editing--trigger-link,
+
+  .inplace-editing--trigger-link
+    color: #888
+    font-size: 1rem
```
This creates a deviation for comments:

![image](https://cloud.githubusercontent.com/assets/617519/9846423/57240ad4-5ad2-11e5-903d-088314a108a8.png)

from description

![image](https://cloud.githubusercontent.com/assets/617519/9846429/6279180c-5ad2-11e5-9304-060b0de88303.png)

and watchers

![image](https://cloud.githubusercontent.com/assets/617519/9846444/7b68bd72-5ad2-11e5-903e-af7c9aa10198.png)

Granted, description and watchers where different before but we shouldn't add a third version.

basiszwo replied:
In my opinion the fields are different because one is / gets a text area and the other is a select field. We should return to the question how we want to display those fields later. For watchers and comments I would directly show the form fields.

HDinger replied:
What do you mean with 'show the form fields'? And shouldn't Description and Comment look alike since they are both text areas and Watchers a select field?

oliverguenther replied:
With your reasoning, I think that at least description and comments should look identical.

I took the comment form because it is _outside_ the usual field attribute groups, and thus one can quickly miss it. With the current styling it looks more similar to a regular input field.

ulferts commented on app/services/update_work_package_service.rb:
```diff
@@ -42,10 +42,21 @@ def update
+  ##
+  # Create journal for the given user and note.
+  # Does not change the work package itself.
+  def add_journal(notes = '')
```
How about renaming this to `create_journal` to communicate the saving within the method.

oliverguenther replied:
:ballot_box_with_check:

ulferts commented on frontend/app/services/activity-service.js:
```diff
@@ -26,19 +26,23 @@
+  NotificationsService,
+  ConfigurationService,
+  $timeout,
+  PathHelper){
```
What he said

oliverguenther replied:
I'm going to mark your comments with a :ballot_box_with_check: to mark them as 'addressed' for myself.

:ballot_box_with_check:

ulferts commented on frontend/app/services/activity-service.js:
```diff
@@ -26,19 +26,23 @@
-      return workPackage.links.addComment.fetch(options);
+    createComment: function(workPackage, comment, notify) {
+      return $http({
+        url: URI(workPackage.links.addComment.url()).addSearch('notify', notify).toString(),
```
How about we do this here http://stackoverflow.com/questions/19156002/does-dependency-injection-in-angularjs-only-work-with-angular-objects

> > could I depend on some external library like jQuery using DI?
> 
> Yes, we are doing it like: myModule.constant("jQuery", window.$).

oliverguenther replied:
I can define it like this in `openproject-app.js`:

```
// globals
openprojectApp
  .constant('URI', URI)
  .config([ .... 
```

~~But I'm not yet convinced that this will please the hound.~~ Yes it will due to DI

ulferts commented on frontend/app/services/activity-service.js:
```diff
@@ -51,6 +55,16 @@ module.exports = function(HALAPIResource, $http, PathHelper){
       };
 
       return activity.links.update.fetch(options).then(function(activity){
+        notification = NotificationsService.addSuccess(
```
Please add the missing `var`

ulferts commented on frontend/app/services/activity-service.js:
```diff
@@ -51,6 +55,16 @@ module.exports = function(HALAPIResource, $http, PathHelper){
+        // but only when we're not in accessible mode
+        if (!ConfigurationService.accessibilityModeEnabled()) {
+          $timeout(function() {
+            NotificationsService.remove(notification);
```
Please rely on the `var` added above.

oliverguenther replied:
Removed variable, as we're not addressing the `hide after timeout` any longer in here. 
:ballot_box_with_check:

ulferts commented on frontend/app/services/activity-service.js:
```diff
@@ -51,6 +55,16 @@ module.exports = function(HALAPIResource, $http, PathHelper){
+        if (!ConfigurationService.accessibilityModeEnabled()) {
+          $timeout(function() {
+            NotificationsService.remove(notification);
+          }, 5000);
```
I think this is the wrong place to handle the disappearing behaviour of notifications.

This should be handled globally by the NotificationService/NotificationDirective to create a behaviour that is consistent. As such, I think it is outside the scope of this PR.

oliverguenther replied:
I agree. Removed :ballot_box_with_check:

ulferts commented:
When adding a comment, the focus lost in the sense, that the page scrolls up and tabbing will start at the beginning of the page (the "Overview" link). If the work package is long, I no longer see the comment I just added. This will probably be an accessibility issue as well but this again touches the field of accessibility on the notifications as well so the solution might be out of scope for accessibility.

For now, I'd see to having the focus either on the newly added comment (preferred) or on the "Add comment" field.  

---

> Appended by @oliverguenther:
> Focus is now set with `scrollIntoView(false)` for WP updates, and by setting `$location.hash` for activities, as they're loaded asynchronously outside the WP refreshment.

ulferts commented:
A notification is shown for updating a comment but not for adding a comment. 

> Appended by @oliverguenther:
> :ballot_box_with_check:

ulferts commented:
The comment field shrinks below it's initial size: 

![image](https://cloud.githubusercontent.com/assets/617519/9847005/71e8a10a-5ad6-11e5-88c0-576bdc030829.png)

To reproduce, I added the text typed `#id of existing wp` and hit backspace. The shrinking also happens, if I manually increased the box before which leads to really weird Look&Feel (the feel part)

---

> Appended by @oliverguenther:
> Should be addressed with elastic (couldn't reproduce it with it) and resizing is now set to `none` for elastic textareas.

ulferts commented:
Don't know if this is feature or bug. When I edited a comment, decided to abort the change by clicking the "x" button and then want to change it again, the text box does not contain the current, unchanged value but my input from the first edit attempt:

![image](https://cloud.githubusercontent.com/assets/617519/9847162/52d1ee10-5ad7-11e5-8da6-072cfafc1ede.png)

Only clicking "edit":
![image](https://cloud.githubusercontent.com/assets/617519/9847171/61031e8c-5ad7-11e5-9e43-70a50850c351.png)

---

> Appended by @oliverguenther:
> Comment is now restored to its original value

ulferts commented:
When I 
1) Set an empty description into edit mode:

![image](https://cloud.githubusercontent.com/assets/617519/9847257/fc30857a-5ad7-11e5-9c85-22ebe526e028.png)

2) Add a comment and hit "Save"

![image](https://cloud.githubusercontent.com/assets/617519/9847263/06d249aa-5ad8-11e5-8422-82d7fa2c8008.png)

An error message is displayed:

![image](https://cloud.githubusercontent.com/assets/617519/9847254/f023dcfa-5ad7-11e5-8f56-c651ce6f9ea5.png)

Not only is the error message displayed incorrectly because it is legitimate to have an empty description but the error message itself is wrong as well. We are not trying to save the work package, at least according to the [WP](https://community.openproject.org/work_packages/20208#note-18)

> comments are being submitted seperate, also are other changes to the work package 

However, based on the server logfile, the work package is in fact saved. These are the requests when having another inplace edit field open:

```
Started POST "/api/v3/work_packages/144/form" for 127.0.0.1 at 2015-09-14 13:10:10 +0200
Started PATCH "/api/v3/work_packages/144?notify=false" for 127.0.0.1 at 2015-09-14 13:10:10 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:10:10 +0200
Started POST "/api/v3/work_packages/144/activities?notify" for 127.0.0.1 at 2015-09-14 13:10:10 +0200
Started POST "/api/v3/work_packages/144/form" for 127.0.0.1 at 2015-09-14 13:10:10 +0200
Started GET "/api/v3/work_packages/schemas/3-2" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started GET "/api/v3/work_packages/144/revisions" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started POST "/api/v3/work_packages/144/form" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started GET "/api/v3/work_packages/schemas/3-2" for 127.0.0.1 at 2015-09-14 13:10:11 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/work_packages/144/revisions" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:10:12 +0200
```

vs the requests when no other field is in edit mode:

```
Started POST "/api/v3/work_packages/144/activities?notify=false" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started POST "/api/v3/work_packages/144/form" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started GET "/api/v3/work_packages/schemas/3-2" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started GET "/api/v3/work_packages/144/revisions" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started GET "/api/v3/work_packages/144" for 127.0.0.1 at 2015-09-14 13:09:24 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:09:25 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:09:25 +0200
Started GET "/api/v3/users/1" for 127.0.0.1 at 2015-09-14 13:09:25 +0200
```

On top of that, I am wondering why the 

```
POST "/api/v3/work_packages/144/form"
```

Is necessary on an activity update when the backend is changed so that it does not require an update of the form because the lock version is not changed.

---

> Appended by @oliverguenther:
> Description can now again be submitted empty. This will have to be addressed for other API-required fields, that are all set to `<field required="required">`, and thus MAY NOT be empty.
> Concerning the form, we need to refresh it as it contains comments embedded

ulferts commented on frontend/app/templates/work_packages/activities/_user.html:
```diff
@@ -27,23 +27,34 @@
+            <accessible-by-keyboard execute="updateComment()"
+                                    ng-disabled="editCommentForm.$invalid"
+                                    class="inplace-edit--control inplace-edit--control--save">
+              <icon-wrapper icon-name="yes" icon-title="{{ I18n.t('js.button_save') }}">
```
Have done it wrong in the past myself but [Angular 1.3 supports one way bindings](http://blog.thoughtram.io/angularjs/2014/10/14/exploring-angular-1.3-one-time-bindings.html) as well which should be sufficient here.

oliverguenther replied:
You mean like `icon-title="{{::I18n.t('js.button_save') }}"` ? This is evaluated to "" and removes the label for me.. :confused:

ulferts replied:
Hm, probably something along the lines that I18n.t is not yet initialized when the template is rendered. But this is just guesswork.

ulferts commented on frontend/app/templates/work_packages/activities/_user.html:
```diff
@@ -27,23 +27,34 @@
-                    rows="4"
-                    required>
-          </textarea>
+      <form class="user-comment--form inplace-edit--write-value">
```
Is it correct to use a form element here? This form can never be submitted (no action) and we have no end-point for this accepting form input anyway. 

We have done it at other places as well but I am wondering why. Is there some standard we have to adhere to?

oliverguenther replied:
Inputs [appear to be valid since HTML4](http://www.w3.org/TR/html401/interact/forms.html#form-controls), so I'm guessing there's nothing behind it but nostalgia. I'm removing it here.

ulferts commented on frontend/app/templates/work_packages/activities/_user.html:
```diff
@@ -27,23 +27,34 @@
+                      ng-hide="inPreview"
+                      preview-toggle="toggleCommentPreview()"
+                      ng-model="activity.props.comment.raw"
+                      ng-bind-html="activity.props.comment.html"
```
You are directly binding to the structure of the server's JSON response here. A model layer would remove the dependency.

oliverguenther replied:
I've addressed this alongside the comment of yours regarding the repeated edit containing the previous edit (I've removed this, it always shows the original comment now when starting to edit).

HDinger committed:
Incorporate review feedback

* Notification when comment is created
* Better binding to editing comments
* Quote with inserted text fixed

oliverguenther commented:
> For now, I'd see to having the focus either on the newly added comment (preferred) or on the "Add comment" field.

I agree, but I'm not sure how I can scroll to the given element after the work package has been refreshed. I'm using the callback of `workPackageRefreshRequired` at the moment to set `$location.hash` to the newest inserted comment.

This does work, but as we still load the users of each activity afterwards, their usernames and avatars change the position of the scrollbar as soon as they are loaded. I'm not sure how we can address this.

ulferts commented:
If I start to add a new comment

![image](https://cloud.githubusercontent.com/assets/617519/9851021/ebdc7c1e-5af6-11e5-9962-d1c7cde25ad4.png)

And then click on "Quote" on an already existing comment, the text I have written so far is replaced by the quote:

![image](https://cloud.githubusercontent.com/assets/617519/9851049/11ddec7c-5af7-11e5-9ca0-42de48c8cbd2.png)

We should either disable quoting when editing is in currently being conducted or add the quoted text to the text the user has inserted so far (preferred)

---

> Appended by @oliverguenther:
> Quoted text is now appended (with a newline beforehand, so the quote always starts in a new line) IF there is existing text, otherwise, it is replaced

oliverguenther commented:
> or add the quoted text to the text the user has inserted so far (preferred)

Done (with a newline beforehand, so the quote always starts in a new line)

ulferts commented:
The action items (quote, edit) on a comment are not accessible by keyboard as they only appear when I hover over them:

![image](https://cloud.githubusercontent.com/assets/617519/9851139/728573e2-5af7-11e5-9d31-6da301d3a1cd.png)

When not hovering over the edit field, the icons are invisible and therefore not focusable. I had the same problem on watchers and attachments and solved it by maintaining a variable to see where the focus is. If the focus is on e.g. the #id field, we could also display the action items.

---

> Appended by @oliverguenther:
> Addressed in bad workaround in https://github.com/HDinger/openproject/commit/6bbc755c36839f9d45259250288cef849162e647 due to unclear propagation of the blur event.

ulferts commented on frontend/tests/unit/tests/services/activity-service-test.js:
```diff
@@ -30,50 +30,56 @@
-    afterEach(function() {
-      workPackage.links.addComment.fetch.restore();
-    });
+      apiFetchResource = ActivityService.createComment(
```
`apiFetchResource` needs to be defined with a `var` before here

---

EDIT: has been done.

oliverguenther committed:
Replace autogrow with angular-elastic

angular-elastic provides us with an option to always insert an
additional newline to provide smoother animations.

This also changes animations to CSS rather than `jQuery.animate`.

ulferts commented on frontend/bower.json:
```diff
@@ -28,7 +28,7 @@
     "ui-select": "0xf013/ui-select#c7bef79e24cbeab977635c7a94d8f9504d4ee2e2",
     "mousetrap": "~1.4.6",
-    "ng-file-upload": "~5.0.9"
+    "ng-file-upload": "~5.0.9",
```
@oliverguenther the ", " needs to be removed

ulferts commented on frontend/package.json:
```diff
@@ -43,6 +43,7 @@
   },
   "dependencies": {
     "URIjs": "^1.14.1",
+    "angular-elastic": "^2.5.0",
```
@oliverguenther this is not installed upon running `npm install` and I guess it has to do with shrinkwrap.

oliverguenther replied:
Hmm... when I use `npm shrinkwrap` myself, this updates all sorts of packages. Probably not what I mean to do, do you know how that crazy new stuff works?

ulferts replied:
Please take a look here: https://github.com/opf/openproject/blob/dev/frontend/doc/README.md#changing-or-updating-dependencies

oliverguenther committed:
Scroll to textarea controls when editing and space available

ulferts commented:
The "Add comment"-field is more narrow than the description field. It does not span the whole width of the details pane like the description does.

![image](https://cloud.githubusercontent.com/assets/617519/9873229/9fc382f2-5b9e-11e5-8ca3-48ed3944c9c8.png)

![image](https://cloud.githubusercontent.com/assets/617519/9873236/a93f93e8-5b9e-11e5-83bf-eb956447d9b5.png)

---

> Appended by @oliverguenther:
> Addressed alongside consistent border styling in https://github.com/HDinger/openproject/commit/0a07238e465d157486580ba815ae13d0f0867083

oliverguenther committed:
Consistent styling of description, watcher and comment

This re-applies the input styling for comment, and makes it
consistent across description and watcher fields.

oliverguenther committed:
Restore empty description submit

oliverguenther committed:
Workaround for focus event on user activity

It seems that the user activity doesn't correctly handle blur events.
I do not know why, but to provide a workaround, I've forced setting the
focus for the quote and edit buttons.

oliverguenther committed:
Use promises in submitFields methods to defer updating

As submitting comments and fields in one go requires
work packages to be submitted first, we can use promises
to ease the processing of each and abort early, e.g.,
when the WP couldn't be updated, we don't want to submit the comment.

oliverguenther committed:
Allow focussing an element after work package has been updated

This commit allows comments or fields to specify a callback after
the work package has been updated.

This is currently used to scroll to the element (such as the inserted comment),
whenever the work package has loaded.

oliverguenther committed:
Minor corrections

This commit addresses the following minor corrections from review:

1. Remove constant URI:
It breaks karma, and I can't really tell why

2. Avoid using Object.values

Appeared to be supported by chrome, but causes the protractor failure.
This also switches the ordering to binary (unshift, push) as that's all
we need for now.

3. Remove location.hash when jumping to element

`$location.hash` may take precedence over `scrollIntoView` ocassionaly,
thus we forcefully remove the location hash when updating the work
package on a particular field

4. Avoid showing comment success when empty

When opening the comment field, and then submitting on another
field with comment being empty and open, the comment was not submitted,
but a successful comment notification was still sent.

5. Fix issues for IE 10+

* Force min-height to be set on the element, as IE tends to return 0 as
  height when empty

6. Misc.

* Avoid undefined in comment text
* Avoid resizing [msd-elastic] textareas
* Rename `add_journal` to `create_journal`

Correct wrong bracket for reject function

oliverguenther commented:
I've just tested this on IE 10/Edge:
1. IE seems to return an invalid height for `angular-elastic`, so that field collapses. Known bug: https://github.com/monospaced/angular-elastic/issues/58. I've addressed this by forcing a min-height style on the affected textareas.
2. The icons are misplaced, however that happens for FF and on dev as well. We (@HDinger and I) will address this in the restyling of these icons.

![bildschirmfoto 2015-09-16 um 15 18 45](https://cloud.githubusercontent.com/assets/459462/9905620/50c1b0e6-5c86-11e5-89bc-e17bc4a62882.png)

basiszwo commented:
LGTM, waiting for travis

ulferts commented on frontend/app/work_packages/activities/user-activity-directive.js:
```diff
@@ -99,20 +104,37 @@ module.exports = function($uiViewScroll,
+
+          }), function() {
+            this.inPreview = false;
+          };
```
I believe that houndci spotted an error here. Instead of having written

```
then(function(r) {
...
}, function() { 
...
});
```

where the second function becomes the `reject` callback of the promise, you have misplaced the closing bracket which will render the intended `reject` callback useless.

ulferts commented on frontend/app/work_packages/controllers/work-package-details-controller.js:
```diff
@@ -71,6 +72,7 @@ module.exports = function($scope,
       });
   }
   $scope.refreshWorkPackage = refreshWorkPackage; // expose to child controllers
+  $rootScope.refreshWorkPackage = refreshWorkPackage; // expose to child controllers
```
Why are you exposing the function when the event is already exposed? (`workPackageRefreshRequired`). It does also take a callback so I see no reason to expose the function itself here.

ulferts commented on frontend/app/work_packages/directives/inplace_editor/custom/editable/inplace-editor-wiki-textarea-directive.js:
```diff
@@ -61,13 +60,33 @@ module.exports = function(TextileService, EditableFieldsState, $sce, AutoComplet
+          var controls = element.closest('.inplace-edit--form ')
+                                .find('.inplace-edit--controls');
+
+          controls[0].scrollIntoView(false);
```
Can we prevent scrolling if the control is already within the viewport? Otherwise the position of the comment field will be changed and I have to find the field again.

ulferts commented:
Probably out of scope but noting it down so you have heard about it @og. It is currently possible to send an empty update via the client. This is no big problem as it will not result in any data changes (as far as I have seen) but the client will show a "Successful update" notification without having changed anything.

ulferts commented:
I am done with my review, only some minor annoyances to get rid of before merging!

oliverguenther committed:
Scroll to element only when visible

Introduces a simple helper plugin to determine whether an element is
`visible()`.

oliverguenther commented:
@ulferts I've addressed the scrolling in https://github.com/opf/openproject/commit/d5a3b210497f71f7f599f5ca8fd2d0b24ce43791.
